### ISSUE TITLE
Huge rocketchat_sessions collection and contains old sessions

### ISSUE THREAD

**Issue by emikolajczak** [RocketChat/Rocket.Chat#17117]
### Description:

I have noticed that `rocketchat_sessions` collection is very huge, about `880894` entries and majority of it has set `closedAt`. On our server we have about 5000 users. Is this collection should be cleaned? Size of this collection might cause paerformance issues. In Mongo logs I see a lot of entries about long time queries. I.e:

`2020-04-01T07:11:57.294+0200 I COMMAND  [conn50118] command rocketchat.rocketchat_sessions command: aggregate { aggregate: "rocketchat_sessions", pipeline: [ { $match: { type: "user_daily", devices.device.os.name: { $exists: true }, year: 2020, month: 3, day: { $gte: 1, $lte: 31 } } }, { $unwind: "$devices" }, { $group: { _id: { name: "$devices.device.os.name", version: "$devices.device.os.version" }, count: { $sum: "$devices.sessions" }, time: { $sum: "$devices.time" } } }, { $project: { _id: 0, name: "$_id.name", version: "$_id.version", count: 1, time: 1 } } ], allowDiskUse: true, cursor: {}, $clusterTime: { clusterTime: Timestamp(1585717916, 3), signature: { hash: BinData(0, F6A2EC874C1DB86D0C5C100134CCF0AB886F37B2), keyId: 6778075279240724481 } }, $db: "rocketchat" } planSummary: IXSCAN { year: 1, month: 1, day: 1, type: 1 } keysExamined:47599 docsExamined:47568 cursorExhausted:1 numYields:378 nreturned:60 reslen:4115 locks:{ Global: { acquireCount: { r: 392 } }, Database: { acquireCount: { r: 392 } }, Collection: { acquireCount: { r: 392 } } } storage:{} protocol:op_msg 535ms`

### Steps to reproduce:

1. Check size of `rocketchat_sessions` collection
2. Check Mongo logs for long queries.

### Expected behavior:

In my opinion old sessions should be removed from `rocketchat_sessions` collection. Also please advice what can be removed manualy from this collection.

### Actual behavior:

`rocketchat_sessions` is not cleaned.

### Server Setup Information:

- Version of Rocket.Chat Server: 3.0.5
- Operating System: Centos 7
- Deployment Method: docker compose
- Number of Running Instances: 30
- DB Replicaset Oplog: YES
- NodeJS Version: v12.14.0
- MongoDB Version: 4.0.14

**Comment by rodrigok** [comment]
Hi @emikolajczak 

This collection exists to track the users' activity, we have a [index there](https://github.com/RocketChat/Rocket.Chat/blob/develop/app/models/server/models/Sessions.js#L381) to remove the records from the database after 45 days from the computed date, so we keep just a few records for Daily, Monthly and Yearly data.

Since you have 5k users, you can expect, at least, 1 session a day, for 45 days it's expected to have  225k records there, but usually users initiate more than 1 session a day so it probably will be more than 225k as you see in your database.

Those queries should run only once a day as [you can see here](https://github.com/RocketChat/Rocket.Chat/blob/develop/app/statistics/server/lib/SAUMonitor.js#L316) and they may be a bit slow but should not affect your performance since we execute it one once a day.

Does it answer your questions?

Thanks

**Comment by emikolajczak** [comment]
Hi, thank you for explanation.
Additional question, what is cause of store users sessions data for 45 days? Could this be managed by adminstarative options?
I am woried about that size odf this collection in short time exceed size of messages.

**Comment by rodrigok** [comment]
@emikolajczak it's 45 days to ensure we have the latest month data do process the monthly records. Since it's a database index it can't be a setting on admin, but you can reduce it on the database to 31 days if you need.

For instance, on our open server (open.rocket.chat) we have 732.803 records right now and we usually have 2k users a day.

This size should not be a problem and the data will not grow indefinitely since the index will keep removing them, but you can follow it and let us know if it increases more.

Thanks

**Comment by emikolajczak** [comment]
For wich feature this functionality is needed? Which monthly records?
Ok, I will monitor this on our database and let you know.

**Comment by rodrigok** [comment]
@emikolajczak you can check that data on you `Admin > Info` panel, It reports the data to us as well, if you have the statistics report enabled and it fills up the `Engagement Dashboard` feature that we are releasing on version 3.1 (enterprise only). It will be used for some reports in the future as well, so we introduced the data collector before to have the info in place when launching those new features.

**Comment by emikolajczak** [comment]
Ok, so maybe this should be Administrator decision if want to collect this data or not? This is only my proposition :)

**Comment by rodrigok** [comment]
We have a [PR](https://github.com/RocketChat/Rocket.Chat/pull/17114) adding a Troubleshoot section that will allow administrators to disable some features, this is one of them. It will land on version 3.1 so you can disable it if you want.

**Comment by emikolajczak** [comment]
Hi,
I have checked and yesterday `rocketchat_sessions` increased by 25000 entries.
I have performed tests and noticed strange behavior, every browser page reload, when rocketchat is open, insert new item for this user into collection, and sed closedDate in previous. 

`{
        "_id" : "hYxTN6TpCLWvvEHby",
        "day" : 2,
        "instanceId" : "pbmbKhoPPgb2JkpDR",
        "month" : 4,
        "sessionId" : "PMYzTwsam35eMmTAd",
        "year" : 2020,
        "_updatedAt" : ISODate("2020-04-02T10:15:43.224Z"),
        "createdAt" : ISODate("2020-04-02T10:15:40.624Z"),
        "device" : {
                "type" : "desktop-app",
                "os" : {
                        "name" : "Windows",
                        "version" : "10"
                },
                "name" : "Rocket.Chat",
                "longVersion" : "2.17.7",
                "version" : "2.17.7"
        },
        "host" : "chat-dev.czk.comarch.com",
        "ip" : "89.64.34.8",
        "loginAt" : ISODate("2020-04-02T10:15:40.609Z"),
        "type" : "session",
        "userId" : "MKTW8XAYmoRx95YkZ",
        "closedAt" : ISODate("2020-04-02T10:15:43.224Z"),
        "lastActivityAt" : ISODate("2020-04-02T10:15:43.224Z")
}
{
        "_id" : "ewJqtEaovjcJDAZZi",
        "day" : 2,
        "instanceId" : "pbmbKhoPPgb2JkpDR",
        "month" : 4,
        "sessionId" : "wzimYCjiGb62MLrLZ",
        "year" : 2020,
        "_updatedAt" : ISODate("2020-04-02T10:15:44.952Z"),
        "createdAt" : ISODate("2020-04-02T10:15:44.952Z"),
        "device" : {
                "type" : "desktop-app",
                "os" : {
                        "name" : "Windows",
                        "version" : "10"
                },
                "name" : "Rocket.Chat",
                "longVersion" : "2.17.7",
                "version" : "2.17.7"
        },
        "host" : "chat-dev.czk.comarch.com",
        "ip" : "89.64.34.8",
        "loginAt" : ISODate("2020-04-02T10:15:44.952Z"),
        "type" : "session",
        "userId" : "MKTW8XAYmoRx95YkZ"
}`

Could you check this behaviour?

**Comment by 564064202** [comment]
Sir,please give us your docker-compose.yml file,and we can try to analyze it

**Comment by emikolajczak** [comment]
> @emikolajczak it's 45 days to ensure we have the latest month data do process the monthly records. Since it's a database index it can't be a setting on admin, but you can reduce it on the database to 31 days if you need.
> 
> For instance, on our open server (open.rocket.chat) we have 732.803 records right now and we usually have 2k users a day.
> 
> This size should not be a problem and the data will not grow indefinitely since the index will keep removing them, but you can follow it and let us know if it increases more.
> 
> Thanks

Hi @rodrigok , do you see any issue, except statistics, when I set this index to 10 days?

**Comment by rodrigok** [comment]
Since the [PR](https://github.com/RocketChat/Rocket.Chat/pull/17114) added a setting under the Troubleshoot section allowing administrators to disable this feature since version 3.1 I'll move on and close this issue.

Please open a new issue if you have any suggestions on how to improve the behavior of the statistics.

Thanks
